### PR TITLE
Allow User Supplied Price Upcharge

### LINKED ISSUE #373: SI: User Supplied Price Upcharge
## Is your feature request related to a problem?
When using user-supplied price products for donation-style products, it's very common for non-profit organizations to offer for customers to cover the credit card transaction fee.  This is often presented using a radio button or checkbox in some websites.  

A non-profit store owner would like to enable a similar feature in Hotcakes Commerce to allow for a specified upcharge to be applied to the order when this option is selected.  

## Describe the solution you'd like  
Two new fields should be added to the product details view.  They are listed below.  

- **Allow Upcharge?**  (checkbox, default to unchecked)  
- **Upcharge Amount** (Use the same textbox and dropdown list combo option seen in promotions where a number and amount/percentage are seen.  This field will only be visible and editable if the previous checkbox is checked.  The default value for this field should be 3%.  There should only be two possible choices, so front-end validation is not necessary. 

When this new setting is enabled and a valid value is found for the upcharge amount, this amount is applied as a new line item in the cart.  

- This line item should share some of the same behavior as the Free Item logic.  
- The upcharge line item can't be deleted.
- The upcharge line item quantity can't be changed.  
- If the donation product is deleted, so is the upcharge line item.  
- If the donation product is added without the "Cover Credit Card Fees" checkbox being checked, this additional line item shouldn't exist.  

## Describe alternatives you've considered
None at this time.  

## Additional context
Customer support request 7671

### PR DESCRIPTION
**Pull request by ArielBlanco1990**
## Related to Issue
Fixes #373 

## Description

This update contains the changes necessary to add the new "User Supplied Price Upcharge" functionality.

## How Has This Been Tested?

This update was tested on DNN 9.11.0 and was also tested using the upgrade process on an existing solution in a previous version of HCC

## Screenshots (if appropriate):
https://app.screencast.com/w9JOAwJWww2mY

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: Libraries/Hotcakes.Commerce/Orders/LineItem.cs [hunk 1]
```diff
@@ -590,6 +666,17 @@ public string DiscountDetailsAsHtml()
...
         }
 
+        /// <summary>
+        ///     Generates a line by line summary of the upcharge that are in this line item.
+        /// </summary>
+        /// <returns>String - an HTML version of the line items for display to the merchant.</returns>
+        public string UpchargeDetailsAsHtml()
+        {
+            var sb = new StringBuilder();
+            sb.Append($"Upcharge Amount: ({TotalUpcharge():C})<br />");
```
Comment (WillStrohl) [inline Libraries/Hotcakes.Commerce/Orders/LineItem.cs:676]: @ArielBlanco1990 This needs to be localized.

File: Website/DesktopModules/Hotcakes/Core/Controllers/ProductsController.cs [hunk 2]
```diff
@@ -834,6 +834,15 @@ private LineItem ConvertProductToLineItem(ProductPageViewModel model)
 
             var li = model.LocalProduct.ConvertToLineItem(HccApp, model.Quantity, model.Selections, userPrice);
 
+            if (model.LocalProduct.AllowUpcharge)
+            {
+                var boolCoverCreditCardFees = Request.Form["covercreditcard"] == "on";
```
Comment (WillStrohl) [inline Website/DesktopModules/Hotcakes/Core/Controllers/ProductsController.cs:839]: @ArielBlanco1990 Please give this control another selector name, such as `upchargeallowed` or something.  Leave the term "credit card" out of it.  That is only one possible use case for this feature.  😊

File: Website/DesktopModules/Hotcakes/Hotcakes.Modules.csproj [hunk 3]
```diff
@@ -3789,6 +3789,7 @@
     </Content>
     <Content Include="Providers\DataProviders\SqlDataProvider\03.06.00.SqlDataProvider" />
     <Content Include="Providers\DataProviders\SqlDataProvider\03.09.00.SqlDataProvider" />
+    <Content Include="Providers\DataProviders\SqlDataProvider\03.09.01.SqlDataProvider" />
```
Comment (WillStrohl) [inline Website/DesktopModules/Hotcakes/Hotcakes.Modules.csproj:3792]: @ArielBlanco1990 This update is going to be for the 3.9.0 release, not 3.9.1.

File: Website/DesktopModules/Hotcakes/Hotcakes.dnn [hunk 4]
```diff
@@ -58,7 +58,7 @@
 						<attributes>
 							<businessControllerClass>Hotcakes.Modules.Core.HotcakesController, Hotcakes.Modules</businessControllerClass>
 							<desktopModuleID>[DESKTOPMODULEID]</desktopModuleID>
-							<upgradeVersionsList>01.00.00,01.00.07,01.05.00,01.06.00,01.07.00,01.08.00,01.09.00,01.10.00,02.00.00,02.00.01,02.00.02,02.00.03,02.00.04,02.00.05,02.00.06,03.00.00,03.00.01,03.01.00,03.02.00,03.02.01,03.02.02,03.02.03,03.03.00,03.04.00,03.04.01,03.05.00,03.06.00,03.07.00,03.08.00,03.08.01,03.08.02,03.09.00</upgradeVersionsList>
+							<upgradeVersionsList>01.00.00,01.00.07,01.05.00,01.06.00,01.07.00,01.08.00,01.09.00,01.10.00,02.00.00,02.00.01,02.00.02,02.00.03,02.00.04,02.00.05,02.00.06,03.00.00,03.00.01,03.01.00,03.02.00,03.02.01,03.02.02,03.02.03,03.03.00,03.04.00,03.04.01,03.05.00,03.06.00,03.07.00,03.08.00,03.08.01,03.08.02,03.09.00,03.09.01</upgradeVersionsList>
```
Comment (WillStrohl) [inline Website/DesktopModules/Hotcakes/Hotcakes.dnn:61]: @ArielBlanco1990 This update is going to be for the 3.9.0 release, not 3.9.1.

File: Website/DesktopModules/Hotcakes/Hotcakes.dnn [hunk 5]
```diff
@@ -215,6 +215,11 @@
 							<name>03.09.00.SqlDataProvider</name>
 							<version>03.09.00</version>
 						</script>
+            <script type="Install">
+							<path>Providers\DataProviders\SqlDataProvider</path>
+							<name>03.09.01.SqlDataProvider</name>
+							<version>03.09.01</version>
```
Comment (WillStrohl) [inline Website/DesktopModules/Hotcakes/Hotcakes.dnn:221]: @ArielBlanco1990 This update is going to be for the 3.9.0 release, not 3.9.1.

File: Website/DesktopModules/Hotcakes/Providers/DataProviders/SqlDataProvider/03.09.01.SqlDataProvider [hunk 6]
```diff
@@ -0,0 +1,15 @@
+﻿ALTER TABLE {databaseOwner}[{objectQualifier}hcc_Product] ADD
```
Comment (WillStrohl) [inline Website/DesktopModules/Hotcakes/Providers/DataProviders/SqlDataProvider/03.09.01.SqlDataProvider:1]: @ArielBlanco1990 When we write these scripts, they should be written in such a way that it assumes someone might run it a second time.  So, these statements should each first check to see if the column exists.  If it doesn't exist, create it.

**Review by WillStrohl — CHANGES_REQUESTED** [review]
@ArielBlanco1990 There are a few things to address in this PR.  Thanks!  😎

File: Website/DesktopModules/Hotcakes/Core/Controllers/ProductsController.cs (hunk 2 above)
Comment (WillStrohl) [reply Website/DesktopModules/Hotcakes/Core/Controllers/ProductsController.cs:839]: @ArielBlanco1990 This should be done in the viewsets as well.

File: Libraries/Hotcakes.Commerce/Orders/LineItem.cs (hunk 1 above)
Comment (ArielBlanco1990) [reply Libraries/Hotcakes.Commerce/Orders/LineItem.cs]: @WillStrohl This is now resolved.

File: Website/DesktopModules/Hotcakes/Core/Controllers/ProductsController.cs (hunk 2 above)
Comment (ArielBlanco1990) [reply Website/DesktopModules/Hotcakes/Core/Controllers/ProductsController.cs]: @WillStrohl This is now resolved.

File: Website/DesktopModules/Hotcakes/Hotcakes.Modules.csproj (hunk 3 above)
Comment (ArielBlanco1990) [reply Website/DesktopModules/Hotcakes/Hotcakes.Modules.csproj]: @WillStrohl This is now resolved.

File: Website/DesktopModules/Hotcakes/Hotcakes.dnn (hunk 4 above)
Comment (ArielBlanco1990) [reply Website/DesktopModules/Hotcakes/Hotcakes.dnn]: @WillStrohl This is now resolved.

File: Website/DesktopModules/Hotcakes/Hotcakes.dnn (hunk 5 above)
Comment (ArielBlanco1990) [reply Website/DesktopModules/Hotcakes/Hotcakes.dnn]: @WillStrohl This is now resolved.

File: Website/DesktopModules/Hotcakes/Providers/DataProviders/SqlDataProvider/03.09.01.SqlDataProvider (hunk 6 above)
Comment (ArielBlanco1990) [reply Website/DesktopModules/Hotcakes/Providers/DataProviders/SqlDataProvider/03.09.01.SqlDataProvider]: @WillStrohl This is now resolved.

**Comment by ArielBlanco1990** [comment]
@WillStrohl FYI: The PR is ready to be reviewed again, I have made the corrections indicated.

**Review by WillStrohl — APPROVED** [review]
Looks good!

**Pull request merged**